src/lj_str.c

```diff
@@ -161,34 +199,101 @@ GCstr *lj_str_new(lua_State *L, const char *str, size_t lenx)
   h ^= b; h -= lj_rol(b, 16);
   /* Check if the string has already been interned. */
   o = gcref(g->strhash[h & g->strmask]);
+#if LUAJIT_SMART_STRINGS
+/* Regular closed-addressing hash table with fill factor 100% has
+ * "average maximum" collision chain near 6 (probability of longer
+ * collision chain is low). Collision chain of length 40 is certain
+ * sign of bad behaviour, so lets react on it.*/
+#define max_collisions 40
```

Comment:
The default "fast" string hash function samples only a few positions in a string, the remaining bytes don't affect the function's result. The function performs well for short strings; however long strings can yield extremely high collision rates.

An adaptive schema was implemented. Two hash functions are used simultaneously. A bucket is picked based on the output of the fast hash function. If an item is to be inserted in a collision chain longer than a certain threshold, another bucket is picked based on the stronger hash function. Since two hash functions are used simultaneously, insert should consider two buckets. The second bucket is often NOT considered thanks to the bloom filter. The filter is rebuilt during GC cycle.

Parameters below were tuned on a set of benchmarks. Max_collisions is also backed by theory: the length of a collision chain in a hash table with the fill factor of 1.0 doesn't exceed 6 on average (assuming uniformly distributed random keys), hence 40 is a clear indication of an anomaly.